core/src/feature/descriptor.rs

```diff
@@ -0,0 +1,66 @@
+//! This module is for internal organization only and is not meant to be exposed.
+
+use crate::feature::Feature;
+use std::ffi::{c_void, CStr};
+
+/// A dummy pointer to be used to represent zero-sized features.
+///
+/// See the `cast_feature_ref` function below for details.
+static DUMMY_FEATURE_POINTER: () = ();
+
+/// An helper functions that casts a raw feature pointer to a Rust reference to the given feature
+/// type.
+/// If the given pointer is null, this function returns `None`.
+///
+/// This function does an additional check to handle the case where the reference is null but the
+/// Feature type is zero-sized (examples include `HardRTCapable` or `InPlaceBroken`).
+/// If that is the case, a reference to a dummy value is returned instead, which is just as invalid
+/// to read but is non-zero at least.
+///
+/// C LV2 hosts legitimately do this because they have literally nowhere to point to, but in Rust
+/// creating a null reference is instant Undefined Behavior even if the reference itself is never
+/// read and/or zero-sized (e.g. because of `Option`).
+#[inline]
+pub unsafe fn cast_feature_ref<'a, T: Feature<'a>>(feature: *const c_void) -> Option<&'a T> {
```

Comment:
This function feels very wrong to me: It is not associated with a type or trait that gives it some context, it just stands alone. Also, you should handle null pointers differently: Instead of returning a valid-but-useless reference, you should have a special value/type that says "Yes, this is the feature you are looking for, but it does not provide a valid pointer". Methods and types depending on this method should be changed accordingly.

Reply:
I agree that I do not like this function, but I am not sure I understand your fix: can you elaborate it a bit more, possibly with a code snippet perhaps?

Reply:
My fix would be a small refactor; I will create a pull request for it later.